Move overloaded_args from FunctionSignature to PythonArgs

colesbury opened the pull request:
This moves the `overloaded_args` field from FunctionSignature to PythonArgs. FunctionSignature is shared by all calls and should be immutable. PythonArgs contains the parsing results for an single call to the PyTorch API. 

I did not measure a difference in performance in the "overrides_benchmark", although I expect there to be a bit more work in the common case. Note that the noise factor for the benchmark is much larger than the differences reported below:

Before:
```
Type tensor had a minimum time of 2.3615360260009766 us and a standard deviation of 0.7833134150132537 us.
Type SubTensor had a minimum time of 10.473251342773438 us and a standard deviation of 0.1973132457351312 us.
Type WithTorchFunction had a minimum time of 5.484819412231445 us and a standard deviation of 0.13305981701705605 us.
Type SubWithTorchFunction had a minimum time of 11.098146438598633 us and a standard deviation of 0.15598918253090233 us.
```
After:
```
Type tensor had a minimum time of 2.2134780883789062 us and a standard deviation of 0.802064489107579 us.
Type SubTensor had a minimum time of 10.625839233398438 us and a standard deviation of 0.15155907021835446 us.
Type WithTorchFunction had a minimum time of 5.520820617675781 us and a standard deviation of 0.23115111980587244 us.
Type SubWithTorchFunction had a minimum time of 11.227846145629883 us and a standard deviation of 0.23032321769278497 us.
```

Fixes #106974

colesbury committed:
Remove dead code

colesbury committed:
Move overloaded_args to PythonArgs

colesbury committed:
Fix doc

colesbury committed:
Fix lint

ezyang commented:
whoopsies, nice catch!

zou3519 commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -432,18 +432,9 @@ auto handle_torch_function(
       torch_api_function.ptr() != nullptr, "torch API function must exist");
-  py::object ret;
   py::tuple args_ = combine_self_args(self, args);
-  // overloaded_args already all have unique types
-  std::vector<py::object> overloaded_types;
-  overloaded_types.reserve(r.signature.overloaded_args.size());
-  for (auto& arg : r.signature.overloaded_args) {
-    overloaded_types.push_back(
-        py::reinterpret_borrow<py::object>((PyObject*)Py_TYPE(arg.ptr())));
-  }
-  py::tuple py_types = py::cast(overloaded_types);
```
This is just dead code, right?

colesbury replied:
Yes

zou3519 commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -179,7 +179,7 @@ struct PYBIND11_EXPORT PythonArgParser {
 };
 
-struct PYBIND11_EXPORT FunctionSignature {
+struct FunctionSignature {
```
nit: can you add what you wrote about FunctionSignature being immutable as a comment here so we don't attempt to do the same thing in the future?

albanD commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1541,15 +1529,20 @@ PythonArgs PythonArgParser::raw_parse(
   if (signatures_.size() == 1) {
     auto& signature = signatures_[0];
-    signature.parse(self, args, kwargs, parsed_args, true);
+    std::vector<py::handle> overloaded_args;
```
py::handle? We don't own these objects here? Who keeps them alive?

colesbury replied:
Two explanations:

1) The args tuple and kwargs dict own references to the arguments. They are valid for at least as long as the function call.
2) This is how it worked before, so if it worked then it will keep working now!

albanD reviewed (approved):
SGTM

colesbury committed:
Add struct-level docs

colesbury commented:
@pytorchbot merge

colesbury committed:
Use std::vector<PyObject*> instead of std::vector<py::handle>

Hopefully, this fixes the Windows build issues. The use of pybind11
types in the struct field has only bought us plenty of linker headaches.

colesbury commented:
@pytorchbot rebase

colesbury commented:
FYI, I've changed `std::vector<py::handle>` to `std::vector<PyObject*>` in `struct PythonArgs`. The use of `py::handle` in the returned struct caused symbol visibility and linker problems that I wasn't able to resolve. (See e.g., https://github.com/pytorch/pytorch/actions/runs/5860835115/job/15892629532).

colesbury commented:
@pytorchbot merge
